### PR TITLE
Extend and correct the ineffective_bit_mask check

### LINKED ISSUE #203: false positives on ineffective_bit_mask
Revisiting `ineffective_bit_mask`, it appears I got my math wrong.

A false positive: `x | 2 >= 5` is currently flagged as equal to `x >= 5`, which is wrong for `x = 4`.

It appears that the *ineffective* part only works if the comparison is a power of two minus one and the bit mask is less than or equal the compared value.

This is because if there is but one zero-bit `b` below the highest bit in the compared value `c`, a mask that is lower than the compared value could still set that bit in `x` so that `x = c - b` becomes a counterexample.

Well, now I've thought through the math, it's actually quite easy to fix.

### PR DESCRIPTION
**Pull request by llogiq**
closes #203 and #197

### CODE REVIEW

File: tests/compile-fail/bit_masks.rs [hunk 1]
```diff
@@ -46,8 +46,12 @@ fn main() {
 fn ineffective() {
     let x = 5;
 
-    x | 1 > 2; //~ERROR ineffective bit mask
+    x | 1 > 3; //~ERROR ineffective bit mask
     x | 1 < 3; //~ERROR ineffective bit mask
```
Comment (birkenfeld) [inline tests/compile-fail/bit_masks.rs]: is this true for x == 2?

File: src/bit_mask.rs [hunk 2]
```diff
@@ -151,18 +155,36 @@ fn check_bit_mask(cx: &Context, bit_op: BinOp_, cmp_op: BinOp_,
...
+
+fn check_ineffective_gt(cx: &Context, span: Span, m: u64, c: u64, op: &str) {
+    if power_of_two(c + 1) && m <= c {
+        span_lint(cx, INEFFECTIVE_BIT_MASK, span, &format!(
+            "ineffective bit mask: `x {} {}` compared to `{}`, is the same as x compared directly",
+            op, m, c));
+    }
+}
+
+fn power_of_two(x: u64) -> bool {
```
Comment (birkenfeld) [inline src/bit_mask.rs]: there's an `is_power_of_two` method on every int type :)

File: tests/compile-fail/bit_masks.rs (hunk 1 above)
Comment (llogiq) [inline tests/compile-fail/bit_masks.rs]: `x | 1 > 3` ↔ `x > 3`, also for `x` = `2`. However, line 50 is actually an error. I'll investigate.

Comment (birkenfeld) [inline tests/compile-fail/bit_masks.rs]: That's the line I tried to comment on...

**Comment by llogiq** [comment]
r?

**Pull request merged**